OsmAnd/src/net/osmand/plus/views/layers/PointLocationLayer.java

```diff
@@ -69,6 +82,206 @@
 	private boolean nm;
 	private boolean locationOutdated;
 
+	// location pin marker
+	private static final int MARKER_ID_MY_LOCATION = 1;
+	private static final int MARKER_ID_NAVIGATION = 2;
+	private MapMarkersCollection markersCollection;
+	private MapMarker myLocationMarker;
+	private MapMarker navigationMarker;
+	private SWIGTYPE_p_void onSurfaceIconKey = null;
+	private SWIGTYPE_p_void onSurfaceHeadingIconKey = null;
+	private boolean markersNeedInvalidate = true;
+	private boolean hasHeading = false;
+	private boolean hasHeadingCached = true;
+	private Location lastKnownLocation;
+	private float lastHeading = 0.0f;
+
+	private enum MarkerState {
+		Stay,
+		Move,
+		None,
+	}
+
+	private MarkerState currentMarkerState = MarkerState.Stay;
+
+	private void setMarkerState(MarkerState markerState) {
+		if (currentMarkerState == markerState) {
+			return;
+		}
+		currentMarkerState = markerState;
+		updateMarkerState();
+	}
+
+	@Override
+	public void updateLocation(Location location) {
+		lastKnownLocation = location;
+		if (view != null && view.hasMapRenderer()) {
+			view.getApplication().runInUIThread(() -> updateMarkerData(lastKnownLocation, lastKnownLocation.getAccuracy(), null), 0);
```

Comment:
> We have runInUIThread method without delay parameter

Removed.

> Why not pass just lastKnownLocation and heading to updateMarkerData?

I decided to avoid creating 3 separate update methods to avoiding of duplication pretty similar (selection of current marker, and related icons) code for each of dedicated update methods - current implementation of `updateMarkerData(..)` looks like shorter and clear IMHO.

Reply:
Discussed.